### DISCUSSION TITLE
Encryption fastboot images howto?

### DISCUSSION THREAD

**Discussion by arpio23** [droidian/porting-guide#9] [General]
I thought if everyone puts his ideas and knowledge in here, we may end up with something we can put in to the wiki.

What I understand, to have fde working I need:

1. Kernel cmdline needs: droidian.lvm.prefer
2. I need a fastboot flashable image lvm patitioned -> https://github.com/droidian/droidian-encryption-service

Do I put the rootfs.img file in to the rootfs partition or extract rootfs.img in to rootfs partition???
I can create image files and do the partitioning but I never tried to flash it :)
 
I mean a backup and trying it out would work too but maybe someone knows how to do it?

**Comment by nitanmarcel** [comment]
The fastboot flashable zip is mainly created by a github action at https://github.com/droidian-images/droidian

You need to fork that and create an apt repository with the adaptation deb and kernel debs in the apt folder.

**Reply by nitanmarcel** [reply]
There's also the possibility to use an external repository but that requires a few modifications.

I did mine for miatoll droidian since I didn't wanted to mess with changing the sources for the non-fastboot-flashable images.

https://github.com/miatoll-linux/droidian

**Reply by nitanmarcel** [reply]
Also in a side note, the fastboot flashables don't come as sistem images but userdata images :)

**Comment by g7** [comment]
I strongly suggest to use https://github.com/droidian-releng/droidian-build-tools 

It provides helpers to bootstrap a new adaptation, and will also clone the droidian-images/droidian repository for you, configuring it automatically for image builds for your specific device, and creating an internal repository there.

**Reply by arpio23** [reply]
Ok, it finally booting. But I need to place vendor.img in to rootfs https://github.com/arpio23/rootfs-templates/commit/46fe4f5608afb4006c727a32421413da7167616e and mount it in the ramdisk halium script. Is there any way to mount vendor without having 500MB vendor image in rootfs? By the way its a device with A/B slots.

After boot I can see vendor partitions under /dev/disk/by-id/dm-name-dynpart-vendor_a but its not available in ramdisk when halium script is executed.

**Reply by arpio23** [reply]
I cannot test it right now but maybe the line 42 in https://github.com/droidian/lxc-android/blob/bookworm/usr/sbin/mount-android.sh should be:
vendor_images="/userdata/vendor.img /var/lib/lxc/android/vendor.img /dev/mapper/dynpart-vendor$ab_slot_suffix"

**Reply by nitanmarcel** [reply]
> I cannot test it right now but maybe the line 42 in https://github.com/droidian/lxc-android/blob/bookworm/usr/sbin/mount-android.sh should be: vendor_images="/userdata/vendor.img /var/lib/lxc/android/vendor.img /dev/mapper/dynpart-vendor$ab_slot_suffix"

Nope, that's ok.

If you're using the rootfs with encrypted support, the droidian rootfs are included in the userdata.img file, so adding any kind of data on the userdata partition might cause more issues.

**Reply by arpio23** [reply]
> I cannot test it right now but maybe the line 42 in https://github.com/droidian/lxc-android/blob/bookworm/usr/sbin/mount-android.sh should be: vendor_images="/userdata/vendor.img /var/lib/lxc/android/vendor.img /dev/mapper/dynpart-vendor$ab_slot_suffix"

But it works for me with the generic rootfs.img in the userdata partition and I don't need vendor.img in the userdata partition anymore.

Yes it works with the fastboot flashable images too. Don't need the vendor.img anymore. Vendor gets mounted now.

**Reply by g7** [reply]
This is odd, since it shouldn't really matter whether is a fastboot-flashable image or not, so I wonder how it did work on the recovery-flashable rootfs in the first place!

But it's a good idea to add the slotted vendor dynpart as well, will add to lxc-android, thanks.

**Reply by g7** [reply]
lxc-android 1:21 should fix this.

**Reply by arpio23** [reply]
Thansk.

> This is odd, since it shouldn't really matter whether is a fastboot-flashable image or not, so I wonder how it did work on the recovery-flashable rootfs in the first place!
> 
> But it's a good idea to add the slotted vendor dynpart as well, will add to lxc-android, thanks.

On my device it never worked, I always needed the vendor.img in userdata.

**Comment by arpio23** [comment]
Ok, I've got everything build and it boots, but only to ssh.
/vendor is empty.

I've just put a vendor.img in to rootfs and mounted it with systemd service, probably not the right thing but just for testing. And the error still the same, this part keeps looping in dmesg:

`[   23.594052] init: Reboot ending, jumping to kernel
[   23.679399] init: init first stage started!
[   23.679553] init: Skipped setting INIT_AVB_VERSION (not in recovery mode)
[   23.699993] init: init second stage started!
[   23.702126] init: Could not read properties from '/vendor/etc/selinux/vendor_property_contexts': No such file or directory
[   23.702161] init: Could not read properties from '/vendor/etc/selinux/nonplat_property_contexts': No such file or directory
[   23.704725] init: Failed to initialize property area
[   23.713840] init: #00 pc 000000000006af58  /system/bin/init (android::init::InitFatalReboot()+80)
[   23.713866] init: #01 pc 000000000009240c  /system/bin/init (android::init::InitAborter(char const*)+44)
[   23.713880] init: #02 pc 000000000000c5b4  /system/lib64/libbase.so (android::base::LogMessage::~LogMessage()+608)
[   23.713893] init: #03 pc 0000000000059784  /system/bin/init (android::init::property_init()+228)
[   23.713904] init: #04 pc 000000000004c300  /system/bin/init (android::init::SecondStageMain(int, char**)+396)
[   23.713916] init: #05 pc 000000000002318c  /system/bin/init (main+304)
[   23.713926] init: #06 pc 000000000007d78c  /system/lib64/bootstrap/libc.so (__libc_init+108)
`
Same kernel without the droidian.lvm.prefer in the cmdline boots fine with the nightly rootfs.img in userdata.

Attached full bootlog.
[dmesg.txt](https://github.com/droidian/porting-guide/files/10143092/dmesg.txt)

**Comment by arpio23** [comment]
Ok now I have the problem that the unl0kr not showing up. Also plymouth never worked for me, it always shows the Motorola screen until phosh is started. I guess its related issue.

What would be the best way to debug this?

**Reply by nitanmarcel** [reply]
How are you building the kernel? Using docker following the instructions in the porting guide?

https://github.com/droidian/porting-guide/blob/master/kernel-compilation.md#package-bring-up

**Reply by arpio23** [reply]
Yes. And now I've setup github actions. https://github.com/arpio23/kernel_motorola_sm6125/blob/droidian/.github/workflows/work.yml

**Reply by nitanmarcel** [reply]
> Yes. And now I've setup github actions. https://github.com/arpio23/kernel_motorola_sm6125/blob/droidian/.github/workflows/work.yml

Don't reinvent the wheel. You have the exact commands you need to use to build the kernel in the porting documentation.

Install `linux-packaging-snippets`

* `apt-get install linux-packaging-snippets`

(Re)create the `debian/control` file:

* `rm -f debian/control`
* `debian/rules debian/control`

Build the kernel (replace arm64 with your phone's arch:

* `RELENG_HOST_ARCH="arm64" releng-build-package`

**Reply by arpio23** [reply]
Yes, I do it locally but something didn't work in the workflow and I didn't debug it, so just made it work. But for testing I always use my local builds exactly following the porting guide.

But anyway any ideas about plymouth and unl0kr not showing?

**Reply by g7** [reply]
Yes, it's probably the same underlying issue. In Droidian, both plymouth and unl0kr use minui to draw to the framebuffer. minui is typically used in Android recoveries.

Something that can be tried is:

```
sudo su

# Install unl0kr, since it's embedded in the initramfs and thus the package is not required in the rootfs
apt install unl0kr

# Stop phosh and android composer
systemctl stop phosh
systemctl stop android-service@hwcomposer
ps aux | grep composer | grep "vendor" | awk '{print $2 }' | xargs sudo kill -9

# Start unl0kr
unl0kr
```

Paging @berbascum as well since they seem to have a similar issue (https://github.com/droidian/plymouth/issues/2)

**Reply by arpio23** [reply]
So got plymouth working and gonna test unl0kr maybe on the weekend. I had to disable CONFIG_FB_VIRTUAL in the kernel config.

**Reply by berbascum** [reply]
Thanks!

Disabling CONFIG_FB_VIRTUAL seems  not solve the issue on lavender.
